hub/core/query/query.py

```diff
@@ -0,0 +1,280 @@
+from abc import ABC
+from typing import Any, Dict
+
+import hub
+import json
+import numpy as np
+
+from hub.core.tensor import Tensor
+
+
+class DatasetQuery:
+    """Dataset query function which is defined by the query string.
+    Designed not to be used directly, but passed as a function parameter to
+    `dataset.filter()`
+
+    Example:
+
+    >>> query = DatasetQuery(dataset, 'images[1:3].min == 0')
+    >>> query = DatasetQuery(dataset, 'labels > 5')
+    >>> query(sample_in)
+    True
+    """
+
+    def __init__(self, dataset: hub.Dataset, query: str):
+        self._dataset = dataset
+        self._query = query
+        self.global_vars: Dict[str, Any] = dict()
+
+    def __call__(self, *args: Any) -> bool:
+        return self._call_eval(*args)
+
+    def _call_eval(self, sample_in: hub.Dataset):
+        return eval(
+            self._query, {}, self._export_tensors(sample_in)
+        )  # TODO export tensors create new bindings. Can we update instead ?
+
+    def _export_tensors(self, sample_in):
+        bindings = {"dataset": sample_in}
+
+        for key, tensor in sample_in.tensors.items():
+            var = _sanitize_tensor_name(key)
+            if tensor.htype == "class_label":
+                bindings[var] = EvalLabelClassTensor(tensor)
+                for label in tensor.info["class_names"]:
+                    bindings[label] = label
+            elif tensor.htype == "text":
+                bindings[var] = EvalTextTesor(tensor)
+            elif tensor.htype == "json":
+                bindings[var] = EvalJsonTensor(tensor)
+            elif "/" in var:
+                group, tensor_name = var.split("/")
+
+                if not group in bindings:
+                    bindings[group] = EvalGroupTensor(group)
+                bindings[group].extend(tensor_name, tensor)
+
+            else:
+                bindings[var] = EvalGenericTensor(tensor)
+
+        return bindings
+
+    def __repr__(self) -> str:
+        return f"'{self._query}' on {self._dataset}"
+
+
+# TODO very naive sanitize
+def _sanitize_tensor_name(key):
+    """Sanitizes tensorname, that it could be binded as variable for `eval()` function call"""
+    key = key.replace(" ", "_")
+    return key
```

Comment:
Please suggest what sanitization have to be in place  for the tensor bindings,

Reply:
@aliubimov need to loop in UI team for this. I'll follow up in slack.

Reply:
@aliubimov need more context to understand what is this function sanitizing.

Reply:
@activesoull @alphazero this particular function have to ensure, that tensor name could be mapped and used as [Python identifier](https://docs.python.org/3/reference/lexical_analysis.html#identifiers). We can also try rename tensor (like in a example, where we replace space with underscore), but it is not clear what to do with special symbols.

Reply:
@aliubimov if we must ensure that the tensor name is a valid python identifier then it will make sense to just remove all symbols and simplify the string to ensure it contains only the ASCII part of the allowed characters (a-zA-Z0-9_). Everything must be removed. That's my take.

Reply:
@aliubimov its better to not have any sanitization at all, as sanitizing can lead to conflicts.. Image 2 tensors with names "profit_$" and "profit_€". Simply removing the special chars will lead to conflict and escaping the special chars is bad UX. Tensors that are not valid identifiers can be accessed by ds["tensor name"]. This how its already works in hub - non identifier tensor names can't be accessed with `.` operator.

Reply:
@farizrahman4u Good point. I still think that having space replaced by underscore is beneficial. That would be something that I would try first as a user.

Reply:
@farizrahman4u Removed as per request.